src/stream/gzip.rs

```diff
@@ -111,3 +111,96 @@ fn get_header(level: Compression) -> Bytes {
 
     Bytes::from(header)
 }
+
+#[unsafe_project(Unpin)]
+pub struct DecompressedGzipStream<S: Stream<Item = Result<Bytes>>> {
+    #[pin]
+    inner: S,
+    flushing: bool,
+    input_buffer: Bytes,
+    output_buffer: BytesMut,
+    crc: Crc,
+    header_stripped: bool,
+    decompress: Decompress,
+}
+
+impl<S: Stream<Item = Result<Bytes>>> Stream for DecompressedGzipStream<S> {
+    type Item = Result<Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        const OUTPUT_BUFFER_SIZE: usize = 8_000;
+
+        let this = self.project();
+
+        if this.input_buffer.len() <= 8 {
+            if *this.flushing {
+                // check crc and len in the footer
+                let crc = &this.crc.sum().to_le_bytes()[..];
+                let bytes_read = &this.crc.amount().to_le_bytes()[..];
+                if crc != this.input_buffer.slice(0, 4) {
+                    return Poll::Ready(Some(Err(Error::new(
+                        ErrorKind::InvalidData,
+                        "CRC computed does not match",
+                    ))));
+                } else if bytes_read != this.input_buffer.slice(4, 8) {
+                    return Poll::Ready(Some(Err(Error::new(
+                        ErrorKind::InvalidData,
+                        "amount of bytes read does not match",
+                    ))));
+                }
+                return Poll::Ready(None);
+            } else if let Some(bytes) = ready!(this.inner.poll_next(cx)) {
+                this.input_buffer.extend_from_slice(&bytes?);
+            } else {
+                *this.flushing = true;
+            }
+        }
+
+        if !*this.header_stripped {
+            let header = this.input_buffer.split_to(10);
```

Comment:
This will panic if the first chunk contains less than 10 bytes.

Reply:
Yup, though using the state machine implementation I think I might not have to deal with this problem (as if the chunk isn't large enough to strip out the header I can just keep reading more in)